core/src/main/java/com/linecorp/armeria/server/RoutingPredicate.java

```diff
@@ -0,0 +1,255 @@
+                                          "!".equals(m.group(1)) ? predicate.negate() : predicate);
+        }
+
+        final ParsedComparingPredicate parsed = parse(headersPredicate);
+        final Predicate<HttpHeaders> predicate;
+        if ("=".equals(parsed.comparator)) {
+            predicate = headers -> headers.getAll(parsed.name).stream()
+                                          .anyMatch(parsed.value::equals);
+        } else {
+            assert "!=".equals(parsed.comparator);
+            predicate = headers -> headers.getAll(parsed.name).stream()
+                                          .noneMatch(parsed.value::equals);
+        }
+        return new RoutingPredicate<>(headersPredicate, predicate);
+    }
+
+    static RoutingPredicate<HttpParameters> ofParams(String paramName, Predicate<String> valuePredicate) {
+        return new RoutingPredicate<>(paramName,
+                                      params -> params.getAll(paramName).stream().anyMatch(valuePredicate));
+    }
+
+    @VisibleForTesting
+    static RoutingPredicate<HttpParameters> ofParams(String paramsPredicate) {
+        requireNonNull(paramsPredicate, "paramsPredicate");
+        final Matcher m = CONTAIN_PATTERN.matcher(paramsPredicate);
+        if (m.matches()) {
+            final Predicate<HttpParameters> predicate = params -> params.contains(m.group(2));
+            return new RoutingPredicate<>(paramsPredicate,
+                                          "!".equals(m.group(1)) ? predicate.negate() : predicate);
+        }
+
+        final ParsedComparingPredicate parsed = parse(paramsPredicate);
+        final Predicate<HttpParameters> predicate;
+        if ("=".equals(parsed.comparator)) {
+            predicate = params -> params.getAll(parsed.name).stream()
+                                        .anyMatch(parsed.value::equals);
+        } else {
+            assert "!=".equals(parsed.comparator);
+            predicate = params -> params.getAll(parsed.name).stream()
+                                        .noneMatch(parsed.value::equals);
+        }
+        return new RoutingPredicate<>(paramsPredicate, predicate);
+    }
+
+    private final CharSequence name;
+    private final Predicate<T> delegate;
+
+    RoutingPredicate(CharSequence name, Predicate<T> delegate) {
+        this.name = requireNonNull(name, "name");
+        this.delegate = requireNonNull(delegate, "delegate");
+    }
+
+    /**
+     * Returns the name of this predicate. This may be the predicate that a user specified.
+     */
+    CharSequence name() {
+        return name;
+    }
+
+    /**
+     * Tests the specified {@code t} object.
+     *
+     * @see DefaultRoute where this predicate is evalued
+     */
+    public boolean test(T t) {
+        try {
+            return delegate.test(t);
+        } catch (Throwable cause) {
+            if (Flags.verboseExceptionSampler().isSampled(cause.getClass())) {
```

Comment:
I'm curious whether this usage is okay or not. The reason why I don't want to write log message every time is because the abnormal request may not be intended by our user. It may be from abusing or hacking. Let me know if I need to implement this by other way.

Reply:
Sounds OK to me. Could you add some comment about it?